### DISCUSSION TITLE
What are the 1 and 2 blocks

### DISCUSSION THREAD

**Discussion by Josverl** [micropython/micropython#11997] [Core Development] [answered]
I've been trying to better understand memory operations now that I can visualize them over time, and there is more happening onder the bonnet than I grok. 

There are some docs on me management that I've read a few times, but I can't find a description of the 1/2 blocks. 

What are these exactly? 
They always seem to grow in equal ratio in my observation, but that might just be chance. 

Also a few sanity checks
- Max block size - Is this the size of the largest allocated block, and in what unit?  N x block size? 

- Max free size, is that in bytes, or in blocks? 

- Is the stack ever shrunk if there is ample stack space, and the heap is full? 

```
No. of 1-blocks: 536, 2-blocks: 119, max blk sz: 72, max free sz: 328 

```

**Comment by dlech** [comment]
You will probably find this helpful (talk on the MicroPython garbage collector): https://www.youtube.com/watch?v=H_xq8IYjh2w

**Comment by dlech** [comment]
> * Is the stack ever shrunk if there is ample stack space, and the heap is full?

No. Generally the stack size (or at least a minimum stack size) is hard-coded in the linker file for an embedded port. The stack and heap sizes are not modified at runtime.

**Comment by dlech** [comment, marked as answer]
> What are these exactly?

Blocks are chunks of contiguous of memory in the heap. The size is configured by `MICROPY_BYTES_PER_GC_BLOCK`. The default is the size of 4 pointers, so 16 bytes on 32-bit arch and 32 bytes on 64-bit arch.

1-blocks are blocks memory that has been allocated for an object that requires <= the number of bytes in one block. Small objects like `floats` fit in to one block. 2-blocks are two contiguous blocks which are allocated for object that are too big to fit in one block but small enough to fit in two blocks.

**Reply by Josverl** [reply]
Thanks,
A few followups

So the headerblock-tailblock(s) construct are for things requiring 3 or more blocks?
Showing like `S===` or `T==` in the memory map I assume 

Is there a way to determine block size/bitness runtime. 
I'm think I remember bitness, but can't recall where

**Reply by jimmo** [reply]
Yep, the `=` shows the tail blocks for the preceding head block.

> Is there a way to determine block size/bitness runtime.

At runtime? i.e. determine the size of a given object e.g. at the REPL? Not that I'm aware of...

If you know the struct that the object is implemented with, then you can use sizeof. I often use `gdb` on the elf file to get this quickly:

e.g.

```
$ arm-none-eabi-gdb build-PYBV11/firmware.elf 
...
(gdb) print sizeof(mp_obj_float_t)
$1 = 8
```

(i.e. half a GC block on 32-bit systems).

Not sure what you mean by "bitness" sorry?

> So the headerblock-tailblock(s) construct are for things requiring 3 or more blocks?

You might also be interested in
https://github.com/micropython/micropython/pull/5195
which gets the histogram of count-of-objects for each number-of-blocks.

**Reply by dlech** [reply]
> I'm think I remember bitness, but can't recall where

If the `uctypes` module is available, you can use it to get the size of a pointer.

**Reply by cwalther** [reply]
Also see [`sys.maxsize`](https://docs.micropython.org/en/latest/library/sys.html#sys.maxsize).

**Reply by Josverl** [reply]
> Also see [`sys.maxsize`](https://docs.micropython.org/en/latest/library/sys.html#sys.maxsize).

Yep, that's the one that slipped my mind

**Comment by dlech** [comment]
> * Max block size - Is this the size of the largest allocated block, and in what unit?  N x block size?
> * Max free size, is that in bytes, or in blocks?

These numbers are all in blocks.  This means the largest object that has been allocated uses 72 contiguous blocks (max block sz) and the largest contiguous region of unallocated blocks is 328 blocks (max free sz).
